File: app/src/main/java/com/centrefx/idcardinfo/IDCardActivity.java
package com.centrefx.idcardinfo;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import java.util.Objects;

public class IDCardActivity extends AppCompatActivity implements View.OnClickListener {

    private int id_type = StringValues.ID_TYPE_NEW;
    private int id_length = 12;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_idcard);

        Bundle b = getIntent().getExtras();
        assert b != null;
        int idType = b.getInt("ID_TYPE");
//        assert idType != null;

        if (idType == StringValues.ID_TYPE_OLD) {
            TextView tv = findViewById(R.id.tx_id_type);
            tv.setText(R.string.title_old_id);
            TextInputLayout edl = findViewById(R.id.edl_id_num);
            edl.setCounterMaxLength(10);
            id_type = StringValues.ID_TYPE_OLD;
            id_length = 10;
        }

        monitorInput(R.id.ed_id_num, id_length, "Limit exceded.");
        monitorInput(R.id.ed_pup_num, 1, "Limit exceded.");

        Button bt = findViewById(R.id.bt_view_data);
        bt.setOnClickListener(this);
    }

    @Override
    public void onClick(View p1) {
        String id_num = getData(R.id.ed_id_num);
        String pp_numS = getData(R.id.ed_pup_num);

        if (!isValid(id_num, id_length) || !isValid(pp_numS, 1)) {
            Log.d("ActivityId", StringValues.INVALID_ID_PP);
            showDialog(StringValues.INVALID_ID_PP);
            return;
        }

        int pp_num = Integer.parseInt(pp_numS);
        String[] result = new String[6];
        IDInfo idn = new IDInfo(id_num, pp_num, id_type);

        if (isNull(idn.getYear()) || isNull(idn.getDateGender()) || isNull(idn.getProvince())) {
            Log.d("ActivityId", StringValues.CALC_ERROR);
            showDialog(StringValues.CALC_ERROR);
            return;
        }

        result[5] = idn.getDateGender();
        String[] r_dateGender = result[5].split(",");
        result[0] = id_num;
        result[1] = idn.getYear() + " - " + r_dateGender[1];
        result[2] = r_dateGender[0];
        result[3] = idn.getProvince();

        if (id_type == StringValues.ID_TYPE_OLD) {
            if (isNull(idn.getCivilStatus())) {
                Log.d("ActivityId", "Invalid Civil Status");
                return;
            } else {
                result[4] = idn.getCivilStatus();
            }
        }

        ResultFragment ar = ResultFragment.newInstance(result);
        FragmentManager fm = getSupportFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(R.id.frag_layout_result, ar, "ActivityResult");
        ft.commit();
    }

    public void showDialog(String msg) {
        FragmentManager fm = getSupportFragmentManager();
        WarningDialog wd = new WarningDialog(msg);
        wd.show(fm, "WarningDialog");
    }

    public void monitorInput(int id, final int size, final String error) {
        final TextInputEditText ed1 = findViewById(id);
        ed1.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (id_type == StringValues.ID_TYPE_OLD) {
                    TextInputEditText ed = findViewById(R.id.ed_id_num);
                    if (s.toString().length() == size - 1) {
                        ed.setInputType(EditorInfo.TYPE_TEXT_FLAG_CAP_CHARACTERS);
                    }
                }
                if (s.toString().length() > size) {
                    ed1.setError(error);
                }
            }
        });
    }

    public String getData(int id) {
        TextInputEditText ed = findViewById(id);
        return Objects.requireNonNull(ed.getText()).toString();
    }

    public boolean isValid(String s, int size) {
        if (isNull(s)) {
            Log.d("ActivityId", "String s is Null.");
            return false;
        }

        if (s.length() == size) {
            try {
                long x;
                if(size == 1){
                    x = Long.parseLong(s);
                }
                else{
                    if (id_type == StringValues.ID_TYPE_OLD) {
                        x = Long.parseLong(s.substring(0, 9));
                    } else {
                        x = Long.parseLong(s);
                    }
                }
                return true;
            } catch (NumberFormatException | StringIndexOutOfBoundsException e) {
                Log.d("ActivityId", "int x is not valid.", new Throwable(e.getMessage()));
                return false;
            }
        }
        return false;
    }

    public boolean isNull(Object o) {
        return o == null;
    }
}

File: app/src/main/java/com/centrefx/idcardinfo_k/IDCardActivity.kt
package com.centrefx.idcardinfo_k

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AppCompatActivity
import com.centrefx.idcardinfo_k.databinding.ActivityIdcardBinding
import com.centrefx.idcardinfo_k.model.IDInfo
import com.centrefx.idcardinfo_k.model.Result
import com.centrefx.idcardinfo_k.model.Values
import com.google.android.material.textfield.TextInputEditText

class IDCardActivity : AppCompatActivity(), View.OnClickListener {
    private lateinit var binding: ActivityIdcardBinding
    private var idType: Int = 0
    private var idLength: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityIdcardBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val b: Bundle? = intent.extras
        idType = b?.getInt("ID_TYPE")!!

        when (idType) {
            Values.ID_TYPE_OLD -> initView(R.string.title_old_id, Values.ID_OLD_LENGTH)
            Values.ID_TYPE_NEW -> initView(R.string.title_new_id, Values.ID_NEW_LENGTH)
        }
        monitorInput(binding.edIdNum, idLength, "Limit exceeded.")
        monitorInput(binding.edPupNum, 1, "Limit exceeded.")

        binding.btViewData.setOnClickListener(this)

    }

    private fun initView(titleResID: Int, length: Int) {
        val tv = binding.txIdType
        tv.text = getString(titleResID)
        val ed = binding.edlIdNum
        ed.counterMaxLength = length
        idLength = length
    }

    override fun onClick(v: View?) {
        val idNum = binding.edIdNum.text.toString()
        val ppNum = binding.edPupNum.text.toString()

        if (isNotValid(idNum, idLength) || isNotValid(ppNum, 1)) {
            showDialog(Values.INVALID_ID_PP)
            return
        }
        val idn = IDInfo(idNum, ppNum.toInt(), idType)

        if (idn.getDateOfBirth().isEmpty() || idn.gender.isEmpty() || idn.getProvince().isEmpty()) {
            showDialog(Values.CALC_ERROR)
            return
        }

        val result = Result(idNum, idn.getDateOfBirth(), idn.gender, idn.getProvince())
        if (idType == Values.ID_TYPE_OLD) {
            result.civilState = idn.getCivilStatus()!!
        }

        val rf = ResultFragment.newInstance(result)
        val fm = supportFragmentManager
        val ft = fm.beginTransaction()
        ft.replace(binding.fragLayoutResult.id, rf, "ActivityResult")
        ft.commit()
    }

    private fun showDialog(msg: String) {
        val fm = supportFragmentManager
        val wd = WarningDialog(msg)
        wd.show(fm, "WarningDialog")
    }

    private fun monitorInput(edt: TextInputEditText, size: Int, error: String) {
        edt.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable?) {
                if (idType == Values.ID_TYPE_OLD && edt == binding.edIdNum) {
                    if (s.toString().length == size - 1) {
                        edt.inputType = EditorInfo.TYPE_TEXT_FLAG_CAP_CHARACTERS
                    }
                }
                if (s.toString().length > size) {
                    edt.error = error
                }
            }
        })
    }

    private fun isNotValid(s: String, size: Int): Boolean {
        if (s.isNullOrBlank()) {
            return true
        }
        if (s.length == size) {
            try {
                val x: Long = if (size == 1) {
                    s.toLong()
                } else {
                    if (idType == Values.ID_TYPE_OLD) {
                        s.substring(0, 9).toLong()
                    } else {
                        s.toLong()
                    }
                }
                return false
            } catch (e: NumberFormatException) {
                return true
            } catch (e: StringIndexOutOfBoundsException) {
                return true
            }
        }
        return true
    }
}